File: csharp-dotnet/src/Architecture.Core/Domain/Events/IDomainEvent.cs
using System;
using System.Collections.Generic;

namespace Architecture.Core.Domain.Events;

/// <summary>
/// Defines the contract for domain events that represent something meaningful that happened in the domain.
/// Domain events enable loose coupling between aggregates and support event-driven architectures.
/// </summary>
public interface IDomainEvent
{
    /// <summary>
    /// Gets the unique identifier for this domain event.
    /// </summary>
    Guid Id { get; }

    /// <summary>
    /// Gets the timestamp when this domain event occurred.
    /// </summary>
    DateTimeOffset OccurredAt { get; }

    /// <summary>
    /// Gets the correlation identifier that tracks related events across service boundaries.
    /// Used to trace a request or business operation through multiple components.
    /// </summary>
    string? CorrelationId { get; }

    /// <summary>
    /// Gets the causation identifier that links this event to the event or command that caused it.
    /// Used to build causal chains and understand event relationships.
    /// </summary>
    string? CausationId { get; }

    /// <summary>
    /// Gets additional contextual metadata associated with this domain event.
    /// Can include user information, system context, or any other relevant data.
    /// </summary>
    IReadOnlyDictionary<string, object> Metadata { get; }
}

File: typescript-nodejs/architecture-core/src/domain/interfaces/i-domain-event.ts
/**
 * Interface for domain events.
 * Domain events represent something that happened in the domain that is of interest to the business.
 * They are used to communicate between different parts of the system.
 */
export interface IDomainEvent {
  /** Unique identifier for the event */
  readonly id: string;

  /** When the event occurred */
  readonly occurredAt: Date;

  /** Correlation ID for tracking related events across service boundaries */
  readonly correlationId: string | undefined;

  /** Causation ID linking this event to the event that caused it */
  readonly causationId: string | undefined;

  /** Additional metadata about the event */
  readonly metadata: Readonly<Record<string, unknown>>;
}